File: 06_reduction/problems/00_reduction_sum/reduction_sum.f90
module integral_mod
contains
  function int_inv_1_x2(a, b, n) result(s)
    real(8), intent(in) :: a, b
    integer(8), intent(in) :: n
    real(8) :: s, dx, x
    integer(8) :: i
    s = 0.0d0
    dx = (b - a) / real(n, 8)
    ! TODO: 下のループを reduction(+:s) を用いて並列化し, 総和の競合を解消せよ.
    do i = 0, n - 1
       x = a + i * dx
       s = s + 1 / (1 + x * x)
    end do
    ! TODO: 上で始めた parallel do 領域を閉じる (!$omp end parallel do).
    s = s * dx
  end function int_inv_1_x2
end module integral_mod

program reduction_sum
  use integral_mod
  character(len=64) :: arg
  real(8) :: a, b, s
  integer(8) :: n
  a = 0.0d0; b = 1.0d0; n = 1000_8 * 1000_8 * 1000_8
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) a
  end if
  if (command_argument_count() >= 2) then
     call get_command_argument(2, arg); read (arg, *) b
  end if
  if (command_argument_count() >= 3) then
     call get_command_argument(3, arg); read (arg, *) n
  end if
  s = int_inv_1_x2(a, b, n)
  print "(a,f0.6)", "s = ", s
end program reduction_sum

File: 06_reduction/problems/00_reduction_sum/reduction_sum.cpp
#include <cstdio>
#include <cstdlib>

double int_inv_1_x2(double a, double b, long n) {
  double s = 0.0;
  double dx = (b - a) / (double)n;
  // TODO: 下のループを reduction(+:s) を用いて並列化し, 総和の競合を解消せよ.
  for (long i = 0; i < n; i++) {
    double x = a + i * dx;
    s += 1 / (1 + x * x);
  }
  return s * dx;
}

int main(int argc, char ** argv) {
  double a = (argc > 1 ? atof(argv[1]) : 0.0);
  double b = (argc > 2 ? atof(argv[2]) : 1.0);
  long n   = (argc > 3 ? atol(argv[3]) : 1000L * 1000L * 1000L);
  double s = int_inv_1_x2(a, b, n);
  printf("s = %f\n", s);
  return 0;
}
